docs/how-does-version-resolution-work-in-maven-and-gradle.md

```diff
@@ -0,0 +1,117 @@
+# How does version resolution work in Maven and Gradle?
+
+The build system that you choose for building your Java code determines which
+version resolution algorithm is used to choose the versions of your
+dependencies. Unfortunately, the two most popular build systems in open source
+(Maven and Gradle) use different version resolution algorithms. They also use
+different names for the concept: Maven calls the process *dependency mediation*,
+while Gradle calls the process *version conflict resolution*. In this article,
+we will use the term *version resolution*. No matter which you choose, you are
+likely to have direct or indirect dependencies that use the other one, so you
+need to understand how both work.
+
+Before we look at Maven and Gradle specifically, we need to understand what
+happens in the version resolution process. The input of this process is a
+dependency graph which reflects the dependencies as specified by each library in
+the graph, which we call an *unresolved dependency graph*. In such a graph, there
+can be multiple versions of each library. The version resolution process walks
+the unresolved dependency graph and decides which version to use for every
+library encountered. Whether or not the build tool produces a graph as an output
```

Comment:
"Whether or not the build tool produces a graph as an output
of this resolution, it is useful to think of this output in terms of a new
graph. We call the graph representation of this output a *resolved dependency
graph*."

I don't buy this. It's clearer to simply go straight to the linear classpath. The intermediate graph is an unnecessary step.

Reply:
How do you describe the output of `mvn dependency:tree`?

Reply:
debugging text printed on the console

Reply:
The text structure is in a tree format (or as best a tree can be represented in text).

When is it used? When debugging dependency conflicts.

Thus, I would argue that thinking of the result of dependency resolution as a tree (or more correctly, a graph) already has precedent, and is the best format. A linear format is harder to take action on in order to fix dependency problems since the input is a tree, and the input is where problems need to be fixed. 

@suztomo do you have thoughts on this?

Reply:
ping @suztomo

Reply:
I need this paragraph. This input tree and output tree are the fundamental part of comparing Gradle and Maven's behavior.